### DISCUSSION TITLE
Open active document in different editor

### DISCUSSION THREAD

**Discussion by timheuer** [microsoft/vscode-discussions#671] [Extension Development QnA] [answered]
I have a custom editor but was curious if a way to provide a way to open the active document in a different editor? Like a toggle to default text editor. I don’t want to implement preview but rather a toggle

**Comment by r3m0t** [comment]
`workbench.action.reopenWithEditor`/`workbench.action.reopenTextEditor`?

**Reply by timheuer** [reply]
Thanks @r3m0t! This works, but only the first time...subsequent calls feel like a no-op...

I'm doing this in my custom editor registration:

```
let openInTextEditorCommand = vscode.commands.registerTextEditorCommand(AppConstants.openInTextEditor, (editor) => {
          vscode.commands.executeCommand('workbench.action.reopenTextEditor', editor?.document?.uri);
        });

        this.context.subscriptions.push(openInTextEditorCommand);
```

It works one time (and even then not reliably...like I have to wait a bit), but then never works again??

**Reply by isidorn** [reply]
Weird. It should just work. Is this still not working? If that is the case, can you provide a small extension.ts that repros this.

**Reply by timheuer** [reply]
Here is where I'm setting it up -- but both my editor/title icons are still showing: https://github.com/timheuer/resx-editor/blob/diffeditor-detection/package.json#L118

**Reply by timheuer** [reply]
Basically I need these conditions...only show the editor/title command WHEN:
- It is a specific resourcetype (resx) **and** my custom editor is not the active editor
**AND**
- never show it when in diff mode regardless of editor

**Reply by isidorn** [reply]
`resourceExtname` context key is only updated based on explorer focus. So it is not updated based on editor title.
Could you try to mimic what our markdown extension is doing https://github.com/microsoft/vscode/blob/main/extensions/markdown-language-features/package.json#L195

So looking at `editorLangId`. And maybe check some other context keys it is using.

**Reply by timheuer** [reply]
That part is the part that is fine actually, it's the combination of NOT wanting to show it when my customer editor is shown and `isInDiffEditor` -- The first condition I outline is fine (showing the menu command only when resource + !myeditor)...but it is showing in diff view -- so I need to also add a condition when it is showing in diff view to ensure it isn't shown.  I can't quite figure out the right `when` clause to handle both/all cases.

Show when:
- resource extension is `resx`
- not my custom editor ID

Don't show when:
- in Diff editor at all regardless of editor

Simply adding the && `!isInDiffEditor` is not working as a condition because it then shows it in cases still I don't want it to.

**Reply by isidorn** [reply]
@ulugbekna can you please help out Tim in his context key battles?

**Reply by ulugbekna** [reply]
> && !isInDiffEditor

There's also `inDiffEditor` context key that you may want to negate as well - https://github.com/microsoft/vscode/issues/186022

Also, keep in mind that `isInDiffEditor` may simply be missing from context, ie is `undefined`, then `!isInDiffEditor` being true

> && !isInDiffEditor is not working as a condition because it then shows it in cases still I don't want it to.

In which particular cases? I like using the command `Inspect Context Keys` that prints all context keys into the developer console (opened via `Toggle Developer Tools` command) and construct your when expression using the state of context keys that I want to capture in my when clause. More: https://code.visualstudio.com/api/references/when-clause-contexts

**Reply by timheuer** [reply]
Interesting, I'm not seeing isInDiffEditor at all in the context keys when my editor is in diff mode...but I do see a diff key I might be able to use (`workbench.editor.sidebysideEditor`) that may work for my needs here.

**Reply by burkeholland** [reply]
@timheuer you are my hero and also did you figure this out?

**Reply by timheuer** [reply, marked as answer]
Not perfectly (but it works) and I didn't end up using the diffeditor check myself.  https://github.com/timheuer/resx-editor/blob/main/package.json#L113
